src/engine.rs

```diff
@@ -197,7 +206,9 @@ impl Engine {
     }
 
     pub fn next_source(&mut self) -> Option<DataSource> {
+        assert!(self.state == EngineState::ReadyToProvide);
         self.await_loop_response();
+        self.state = EngineState::AwaitingCompletion;
```

Comment:
Does `await_loop_response` block? Is it okay to change the state _after_ we call the method, or does that lead to a possible race condition?

Reply:
It does block, but nothing that happens inside it can access or update the current state, so it commutes with assignment.

In general there should be no races because the concurrency model is an entirely synchronous handoff between the two threads: At any given point one of the threads is always waiting on a response from the other.